SQL snippets documentation 12726 (#12795)
metabase/metabase: docs/users-guide/sql-snippets.md

```
from ORDERS 
left join PRODUCTS on PRODUCT_ID = PRODUCT_ID
```

2. **Right-click on the highlighted section**.
3. **Select Save as snippet** to create a snippet. A modal will pop up with the SQL statement you highlighted. 
4. **Edit, name, and describe your snippet**, then click the save button.

In this case, we named the snippet "Orders and Products". The snippet will now be available for anyone to use. Here's what the snippet looks like in the SQL editor:

```
select * 
{{snippet: Orders and Products}};
```

When editing in the SQL editor, you can now start typing `{{snippet:` and Metabase will present autocomplete options for available snippets.

### Snippet menu

![Snippet sidebar and insertion](./images/sql-snippets/snippet_sidebar_and_insertion.gif)

The SQL editor **sidebar** has a **SQL Snippets** menu to list available and archived snippets.

Click on the code icon on the right side of the SQL editor, below the **Data Reference** book icon (learn about your data) and the **Variables** χ icon. Metabase will slide out a sibebar menu that lists available snippets.

From the SQL Snippets menu, you can:

> - **Preview snippets**. Click on the down arrow to the right of a snippet to see a preview of its SQL code. There's also an option to edit the snippet.

Reviewer: ```suggestion
- **Preview snippets**. Click on the down arrow to the right of a snippet to see its description and a preview of its SQL code. There's also an option to edit the snippet.
```

> - **Insert a snippet at point**. Click on a snippet to insert it into your query at the current cursor point.

Reviewer: ```suggestion
- **Insert a snippet**. Click on a snippet's name to insert it into your query at the cursor's current location.
```

- [**Edit a snippet**](#editing-snippets). You can change a snippet's name, description and code.

> - [**Archive and unarchive a snippet**](#archived-snippets). You can archive a snippet, which removes the snippet from the snippet menu and autocomplete options in the SQL editor.

Reviewer: ```suggestion
- [**Archive and unarchive a snippet**](#archived-snippets). From the Edit modal, you can archive a snippet, which removes the snippet from the snippet menu and autocomplete options in the SQL editor.
```

**Editing a snippet's SQL**. Here's where we have to remind you that with great power comes great responsibility. There is one major caveat when editing snippets, worthy of a callout:

> **Caution: if you edit a snippet and include broken code, you will break every question that uses that snippet.** Make sure to test your code before saving it to an existing snippet.

### Archived snippets

**Archiving** snippets can help keep dated or less relevant snippets out of the way. When you archive a snippet, the snippet no longer populates in the snippet autocomplete dropdown, and the snippet will no longer show up in the main list of of snippets in the **SQL editor** sidebar.

Archiving a snippet does not affect any existing queries that use the snippet, so you can safely archive a snippet without impacting any questions.

You can access an archived snippet from the snippet sidebar menu by clicking on the archived button in the bottom left of the sidebar.

Although there is no way to delete a snippet, you can archive and unarchive a snippet at any time.

> 

Reviewer: Might be worth adding a bit here about snippet permissions, like:

### Snippet permissions
Any user who has SQL editor permissions to at least _one_ of your connected databases will be able to view the snippets sidebar, and will be able to create, edit, and archive or unarchive any and all snippets — even ones that might be intended to be used with databases for which the user doesn't have SQL editing access.
